File: (102)Binary_Tree_Level_Order_Traversal/Solution.java
import java.util.LinkedList;
import java.util.List;
import java.util.ArrayList;

public class Solution {
    public static void main(String[] args) {
        // [3,9,20,null,null,15,7]
        TreeNode root = new TreeNode(3);
        TreeNode node1 = new TreeNode(9);
        TreeNode node2 = new TreeNode(20);
        TreeNode node3 = new TreeNode(15);
        TreeNode node4 = new TreeNode(7);
        // TreeNode node5 = new TreeNode(1);
        // TreeNode node6 = new TreeNode(2);
        root.left = node1;
        root.right = node2;
        node2.left = node3;
        node2.right = node4;
        // node4.left = node5;
        // node4.right = node6;
        System.out.println(levelOrder(root));
    }

    public static List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>> answer = new ArrayList<>();
        if (root == null) {
            return answer;
        }
        // Store the values of node at one level
        List<Integer> oneLevel = new ArrayList<>();

        // BFS Queue
        LinkedList<TreeNode> queue = new LinkedList<>();
        // start with root
        queue.offer(root);
        // Insert level boundary
        queue.offer(null);

        while (!queue.isEmpty()) {
            TreeNode current = queue.poll();
            if (current != null) {
                if (current.left != null) {
                    queue.offer(current.left);
                }
                if (current.right != null) {
                    queue.offer(current.right);
                }
                // Collect the value (not TreeNode object) of current level
                oneLevel.add(current.val);
            } else {
                // One level processing completed
                answer.add(oneLevel);
                oneLevel = new ArrayList<>();
                // Insert level boundary for the next level since the child nodes
                // must have been added into the queue by now. Hence when hitting
                // the boundary we can wrap up one level.
                // Except when all nodes has been processed
                if (queue.size() != 0) {
                    queue.offer(null);
                }
            }
        }
        return answer;
    }
}

File: (102)Binary_Tree_Level_Order_Traversal/Solution.py
import Queue


class TreeNode(object):
	"""
	Definition for a binary tree node.
	"""

    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution(object):

    def levelOrder(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        answer = []
        if root is None:
            return answer

        one_level = []

        q = Queue.Queue()
        # Start with root
        q.put(root)
        # Insert level boundary
        q.put(None)

        while not q.empty():
            current = q.get()

            if current is not None:
                # Store/visit current node
                one_level.append(current.val)
                # Process its children
                if current.left is not None:
                    q.put(current.left)
                if current.right is not None:
                    q.put(current.right)
            else:
                # one level completed
                answer.append(one_level)
                one_level = []
                # Insert level boundary for next level
                # Obviously no next level when all nodes are processed
                if q.qsize() != 0:
                    q.put(None)
        return answer

#######################
# Test
#######################
root = TreeNode(3)
node1 = TreeNode(9)
node2 = TreeNode(20)
node3 = TreeNode(15)
node4 = TreeNode(7)
# node5 = new TreeNode(1);
# node6 = new TreeNode(2);
root.left = node1
root.right = node2
node2.left = node3
node2.right = node4
# node4.left = node5;
# node4.right = node6;
print Solution().levelOrder(root)
